metadata/metadata_api_test.go

```diff
@@ -0,0 +1,381 @@
+	assert.NoError(t, err)
+	assert.Equal(t, string(timestampBytes), string(timestamp2Bytes))
+
+}
+
+func TestSignVerify(t *testing.T) {
+	root, err := Root().FromFile(testutils.RepoDir + "/root.json")
+	assert.NoError(t, err)
+
+	// Locate the public keys we need from root
+	assert.NotEmpty(t, root.Signed.Roles[TARGETS].KeyIDs)
+	targetsKeyID := root.Signed.Roles[TARGETS].KeyIDs[0]
+	assert.NotEmpty(t, root.Signed.Roles[SNAPSHOT].KeyIDs)
+	snapshotKeyID := root.Signed.Roles[SNAPSHOT].KeyIDs[0]
+
+	// Load sample metadata (targets) and assert ...
+	targets, err := Targets().FromFile(testutils.RepoDir + "/targets.json")
+	assert.NoError(t, err)
+	data, err := targets.Signed.MarshalJSON()
+	assert.NoError(t, err)
+	sig := getSignatureByKeyID(targets.Signatures, targetsKeyID)
+
+	// ... it has a single existing signature,
+	assert.Equal(t, 1, len(targets.Signatures))
+
+	// ... which is valid for the correct key.
+	targetsKey := root.Signed.Keys[targetsKeyID]
+	targetsPublicKey, err := targetsKey.ToPublicKey()
+	assert.NoError(t, err)
+	targetsHash := crypto.Hash(0)
+	if targetsKey.Type != KeyTypeEd25519 {
+		targetsHash = crypto.SHA256
+	}
+	targetsVerifier, err := signature.LoadVerifier(targetsPublicKey, targetsHash)
+	assert.NoError(t, err)
+	err = targetsVerifier.VerifySignature(bytes.NewReader(sig), bytes.NewReader(data))
+	assert.NoError(t, err)
+
+	snapshotKey := root.Signed.Keys[snapshotKeyID]
+	snapshotPublicKey, err := snapshotKey.ToPublicKey()
+	assert.NoError(t, err)
+	snapshotHash := crypto.Hash(0)
+	if snapshotKey.Type != KeyTypeEd25519 {
+		snapshotHash = crypto.SHA256
+	}
+	snapshotVerifier, err := signature.LoadVerifier(snapshotPublicKey, snapshotHash)
+	assert.NoError(t, err)
+
+	err = snapshotVerifier.VerifySignature(bytes.NewReader(sig), bytes.NewReader(data))
+	assert.ErrorContains(t, err, "crypto/rsa: verification error")
+
+	signer, err := signature.LoadSignerFromPEMFile(testutils.KeystoreDir+"/snapshot_key", crypto.SHA256, cryptoutils.SkipPassword)
+	assert.NoError(t, err)
+	// Append a new signature with the unrelated key and assert that ...
+	snapshotSig, err := targets.Sign(signer)
+	assert.NoError(t, err)
+	// ... there are now two signatures, and
+	assert.Equal(t, 2, len(targets.Signatures))
+	// ... both are valid for the corresponding keys.
+	err = targetsVerifier.VerifySignature(bytes.NewReader(sig), bytes.NewReader(data))
+	assert.NoError(t, err)
+	err = snapshotVerifier.VerifySignature(bytes.NewReader(snapshotSig.Signature), bytes.NewReader(data))
+	assert.NoError(t, err)
+	// ... the returned (appended) signature is for snapshot key
+	assert.Equal(t, snapshotSig.KeyID, snapshotKeyID)
+
+	// Create and assign (don't append) a new signature and assert that ...
+	signer, err = signature.LoadSignerFromPEMFile(testutils.KeystoreDir+"/timestamp_key", crypto.SHA256, cryptoutils.SkipPassword)
+	assert.NoError(t, err)
+
+	// Append a new signature with the unrelated key and assert that ...
```

Comment:
`Sign()` appends by default, except in the use-cases when we explicitly clear the signatures prior calling `Sign()`. We do this in some other places in the tests, but here specifically we don't clear them. I can change to `Add` if you think it will be more clear.